blockdag/reachability.go

```diff
@@ -547,9 +811,276 @@ func (dag *BlockDAG) updateReachability(node *blockNode, selectedParentAnticone
+
+	modifiedTreeNodes, err = dag.propagateChildIntervals(intervalBeforeReindexRootStart, reindexRootChildNodesBeforeChosen,
+		reindexRootChildNodesBeforeChosenSizes, reindexRootChildNodesBeforeChosenSubtreeSizeMaps)
+	if err != nil {
+		return 0, nil, err
+	}
+	return reindexRootChildNodesBeforeChosenSizesSum, modifiedTreeNodes, nil
+}
+
+func (dag *BlockDAG) tightenReachabilityTreeIntervalsAfterChosenReindexRootChild(
+	reindexRoot *reachabilityTreeNode, reindexRootChildNodesAfterChosen []*reachabilityTreeNode) (
+	reindexRootChildNodesAfterChosenSizesSum uint64, modifiedTreeNodes map[*reachabilityTreeNode]struct{}, err error) {
+
+	reindexRootChildNodesAfterChosenSizes, reindexRootChildNodesAfterChosenSubtreeSizeMaps, reindexRootChildNodesAfterChosenSizesSum :=
+		calcReachabilityTreeNodeSizes(reindexRootChildNodesAfterChosen)
+
+	reindexRootEnd := reindexRoot.interval.end
+	targetRangeAfterReindexRootEnd := reindexRootEnd - reachabilityReindexSlack
+	targetRangeAfterReindexRootStart := targetRangeAfterReindexRootEnd - reindexRootChildNodesAfterChosenSizesSum
+	intervalAfterReindexRootEnd := newReachabilityInterval(targetRangeAfterReindexRootStart, targetRangeAfterReindexRootEnd-1)
+
+	modifiedTreeNodes, err = dag.propagateChildIntervals(intervalAfterReindexRootEnd, reindexRootChildNodesAfterChosen,
+		reindexRootChildNodesAfterChosenSizes, reindexRootChildNodesAfterChosenSubtreeSizeMaps)
+	if err != nil {
+		return 0, nil, err
+	}
+	return reindexRootChildNodesAfterChosenSizesSum, modifiedTreeNodes, nil
+}
+
+func (dag *BlockDAG) expandReachabilityTreeIntervalInChosenReindexRootChild(reindexRoot *reachabilityTreeNode,
+	chosenReindexRootChild *reachabilityTreeNode, reindexRootChildNodesBeforeChosenSizesSum uint64,
+	reindexRootChildNodesAfterChosenSizesSum uint64) (map[*reachabilityTreeNode]struct{}, error) {
+
+	modifiedTreeNodes := make(map[*reachabilityTreeNode]struct{})
+
+	reindexRootStart := reindexRoot.interval.start
+	reindexRootEnd := reindexRoot.interval.end
+	targetRangeForReindexRootChildStart :=
+		reindexRootStart + reindexRootChildNodesBeforeChosenSizesSum + reachabilityReindexSlack
+	targetRangeForReindexRootChildEnd :=
+		reindexRootEnd - reindexRootChildNodesAfterChosenSizesSum - reachabilityReindexSlack - 1
+	newReindexRootChildInterval := newReachabilityInterval(targetRangeForReindexRootChildStart, targetRangeForReindexRootChildEnd)
+
+	if targetRangeForReindexRootChildStart > reindexRootStart || targetRangeForReindexRootChildEnd < reindexRootEnd {
+		// New interval doesn't contain the previous one, propagation is required
+		chosenReindexRootChild.interval = newReachabilityInterval(
+			targetRangeForReindexRootChildStart+reachabilityReindexSlack,
+			targetRangeForReindexRootChildEnd-reachabilityReindexSlack-1,
+		)
+		modifiedNodes, err := chosenReindexRootChild.countSubtreesAndPropagateInterval()
```

Comment:
After the propagate call, interval should be set back to:
`		chosenReindexRootChild.interval = newReachabilityInterval(
			targetRangeForReindexRootChildStart,
			targetRangeForReindexRootChildEnd,
		)`
see [here ](https://github.com/kaspanet/pocs/blob/master/ghostdag/block.py#L387) for reference